app/lib/session_encryptor.rb

```diff
@@ -0,0 +1,63 @@
+class SessionEncryptor
+  KEY_LEN = ActiveSupport::MessageEncryptor.key_len
+
+  def initialize(session_signing_key:)
+    @session_signing_key = session_signing_key
+  end
+
+  def encrypt_session(access_token:, refresh_token:)
+    encrypt_string(
+      {
+        access_token: Base64.encode64(access_token),
+        refresh_token: Base64.encode64(refresh_token),
+      }.to_json,
+    )
+  end
+
+  def decrypt_session(ciphertext)
+    return if ciphertext.blank?
+
+    plaintext = decrypt_string(ciphertext)
+    return unless plaintext
+
+    hash = JSON.parse(plaintext)
+
+    {
+      access_token: Base64.decode64(hash["access_token"]),
+      refresh_token: Base64.decode64(hash["refresh_token"]),
+    }
+  end
+
+protected
+
+  attr_reader :session_signing_key
+
+  def encrypt_string(plaintext)
+    salt = generate_salt
+    ciphertext = encryptor(salt).encrypt_and_sign(plaintext)
+    Base64.urlsafe_encode64("#{salt}$$#{ciphertext}")
+  end
+
+  def decrypt_string(ciphertext)
+    bits = Base64.urlsafe_decode64(ciphertext).split("$$")
+    return nil unless bits.length == 2
+
+    salt = bits[0]
+    message = bits[1]
+
+    encryptor(salt).decrypt_and_verify(message)
+  rescue ArgumentError
+    nil
+  rescue ActiveSupport::MessageEncryptor::InvalidMessage
+    nil
+  end
+
+  def encryptor(salt)
+    key = ActiveSupport::KeyGenerator.new(session_signing_key).generate_key(salt, KEY_LEN)
+    ActiveSupport::MessageEncryptor.new key
+  end
+
+  def generate_salt
+    SecureRandom.hex KEY_LEN
+  end
```

Comment:
Is it worth calling this method `salt` and deleting line 36?

Reply:
The local variable is also used on line 37, and we want both to get the same value.  But I'll just inline the method.

Reply:
I was thinking about memoization but didn't mention it in my previous message:
```
def salt
  @salt ||= SecureRandom.hex(KEY_LEN)
end
```

Reply:
That's fair